prov/util/src/util_cq.c

```diff
@@ -38,58 +38,69 @@
 
 #define UTIL_DEF_CQ_SIZE (1024)
 
-/* Caller must hold `cq_lock` */
-int ofi_cq_write_overflow(struct util_cq *cq, void *context, uint64_t flags, size_t len,
-			  void *buf, uint64_t data, uint64_t tag, fi_addr_t src)
+
+/* While the CQ is full, we continue to add new entries to the auxiliary
+ * queue.
+ */
+static void ofi_cq_insert_aux(struct util_cq *cq,
+			      struct util_cq_aux_entry *entry)
+{
+	entry->cq_slot = ofi_cirque_tail(cq->cirq);
+	entry->cq_slot->flags = UTIL_FLAG_AUX;
+	slist_insert_tail(&entry->list_entry, &cq->aux_queue);
+
+	if (!ofi_cirque_isfull(cq->cirq))
+		ofi_cirque_commit(cq->cirq);
+}
+
+/* Caller must hold 'cq lock' */
+int ofi_cq_write_overflow(struct util_cq *cq, void *context, uint64_t flags,
+			  size_t len, void *buf, uint64_t data, uint64_t tag,
+			  fi_addr_t src)
 {
 	struct util_cq_aux_entry *entry;
 
-	assert(ofi_cirque_isfull(cq->cirq));
+	FI_DBG(cq->domain->prov, FI_LOG_CQ, "writing to CQ overflow list\n");
+	assert(ofi_cirque_freecnt(cq->cirq) <= 1);
 
 	if (!(entry = calloc(1, sizeof(*entry))))
 		return -FI_ENOMEM;
 
-	entry->cq_slot = ofi_cirque_tail(cq->cirq);
-	entry->cq_slot->flags |= UTIL_FLAG_OVERFLOW;
-
 	entry->comp.op_context = context;
 	entry->comp.flags = flags;
 	entry->comp.len = len;
 	entry->comp.buf = buf;
 	entry->comp.data = data;
 	entry->comp.tag = tag;
-
+	entry->comp.err = 0;
```

Comment:
Why does this value need to be forced to 0?

Reply:
The code uses the err value to determine if the CQ entry was written for a successful completion or an error.  The current code writes an ERROR flag into the main CQ slot when an error entry is placed on the auxiliary queue.  That flag does not work when the CQ has overflowed.  In the overflow case, successful entries may be written to the auxiliary queue, which all reference the last slot.  If an error completion is also written, it too will reference the last slot, but we can't mark the CQ slot with ERROR, as that assumes all completions referencing that slot are errors.

Reply:
Ah that makes sense, thanks.